internal/gps/source_manager.go

```diff
@@ -578,6 +578,18 @@ func (sm *SourceMgr) InferConstraint(s string, pi ProjectIdentifier) (Constraint
 		return version.Unpair(), nil
 	}
 
+	// Abbreviated git revision? If the string is present, there's a good shot of
+	// this.
+	if present, _ := sm.RevisionPresentIn(pi, Revision(s)); present {
+		srcg, err := sm.srcCoord.getSourceGatewayFor(context.TODO(), pi)
```

Comment:
Instead of dipping into the source gateway, this should probably be another function on the source manager, just like `RevisionPresentIn` which handles getting the value from the source gateway.
https://github.com/golang/dep/blob/ee6fcfcf50aaecc291631324148a44b9b4a792a9/internal/gps/source_manager.go#L436-L448